### PR TITLE
[Early WIP] Gemma3 support.

### PR DESCRIPTION
**Pull request by krammnic**
#### Context
What is the purpose of this PR? Is it to
- [X] add a new feature
- [ ] fix a bug
- [ ] update tests and/or documentation
- [ ] other (please add here)

Please link to any issues this PR addresses.

#### Changelog
What are the changes made in this PR?
* #2484 

#### Test plan
Please make sure to do each of the following if applicable to your PR. If you're unsure about any one of these just ask and we will happily help. We also have a [contributing page](https://github.com/pytorch/torchtune/blob/main/CONTRIBUTING.md) for some guidance on contributing.

- [X] run pre-commit hooks and linters (make sure you've first installed via `pre-commit install`)
- [ ] add [unit tests](https://github.com/pytorch/torchtune/tree/main/tests/torchtune) for any new functionality
- [X] update [docstrings](https://github.com/pytorch/torchtune/tree/main/docs/source) for any new or updated methods or classes
- [X] run unit tests via `pytest tests`
- [ ] run recipe tests via `pytest tests -m integration_test`
- [ ] manually run any new or modified recipes with sufficient proof of correctness
- [ ] include relevant commands and any other artifacts in this summary (pastes of loss curves, eval results, etc.)

#### UX
If your function changed a public API, please add a dummy example of what the user experience will look like when calling it.
Here is a [docstring example](https://github.com/pytorch/torchtune/blob/6a7951f1cdd0b56a9746ef5935106989415f50e3/torchtune/modules/vision_transformer.py#L285)
and a [tutorial example](https://pytorch.org/torchtune/main/tutorials/qat_finetune.html#applying-qat-to-llama3-models)

- [ ] I did not change any public API
- [ ] I have added an example to docs or docstrings

### CODE REVIEW

**Comment by krammnic** [comment]
- [X] Recipes, recipe registry
- [X] Component builders (with all changes according to the tech report)
- [X] Model builders (with all correct params according to the tech report)
- [ ] SigLIP
- [ ] Convert weight
- [ ] Tokenizer? (I assume that it is almost same as in Gemma2)
- [ ] Manual runs

**Comment by krammnic** [comment]
Was able to run full on 1B (without multimodal)

**Comment by felipemello1** [comment]
Hey Mark, great work! As a sanity check, do you think you could compare vs HF for correctness with a larger sentence?

You can follow this as a start: https://gist.github.com/felipemello1/e3f1b1c358e145c7a4d610cf44cca374

**Comment by krammnic** [comment]
> Hey Mark, great work! As a sanity check, do you think you could compare vs HF for correctness with a larger sentence?
> 
> You can follow this as a start: https://gist.github.com/felipemello1/e3f1b1c358e145c7a4d610cf44cca374

Hey Felipe! Yep, sure, it is still WIP until I will be confident (we will do multimodal runs) and some configs will be fixed by Gemma team

**Comment by bzz** [comment]
Curious about 

> some configs will be fixed by Gemma team

any chance it refers to https://github.com/huggingface/transformers/issues/36683 ? 
That seems to have been patched in Transformers

**Comment by krammnic** [comment]
> Curious about
> 
> > some configs will be fixed by Gemma team
> 
> any chance it refers to [huggingface/transformers#36683](https://github.com/huggingface/transformers/issues/36683) ? That seems to have been patched in Transformers

Hey @bzz. Exactly, the issue is very similar. Unfortunately, we require more information on the conversion stage from the config, which is missing in 4b config.

**Review by ebsmothers — COMMENTED** [review]
Thanks for your patience on this one! A couple high-level comments:

1) The whole text-only 1B vs multimodal 4B+ thing is a bit awkward. I feel like in this PR you wrote a bunch of the SigLIP components but didn't really hook them up to anything, and so the actual builders are all text-only. I think that's fine if there's still stuff up in the air (viz. (2)), but wonder if we should do something similar to our other multimodal models: provide `gemma_decoder`, `gemma_vision_encoder`, then for 4B+ we hook into `EarlyFusion` and for 1B we just use the decoder directly.
2) Can you share more on some of the blockers around HF config for 4B+ models you were alluding to? I want to understand how much we should try to hack around things vs just hold off here.

File: torchtune/models/gemma3/_component_builders.py [hunk 1]
```diff
@@ -0,0 +1,392 @@
...
+            kv_cache=None,
+            max_seq_len=max_seq_len,
+            # QK-norm is required
+            k_norm=GemmaRMSNorm(head_dim, eps=norm_eps),
+            q_norm=GemmaRMSNorm(head_dim, eps=norm_eps),
+            attn_dropout=attn_dropout,
+            # perform global only on the each 6 layer, according to the tech-report
+            sliding_window_size=sliding_window_size if (layer_idx % 6) != 0 or layer_idx == 0 else None,
+            # we don't use softcapping in gemma3
+            query_pre_attn_scalar=query_pre_attn_scalar
```
[the remark this answers is not in the document]
Comment (krammnic) [reply torchtune/models/gemma3/_component_builders.py:242]: Hmmm.. Looks like it is unused, yes.

File: torchtune/models/gemma3/_component_builders.py [hunk 2]
```diff
@@ -0,0 +1,392 @@
...
+            # global rope is scaled
+            pos_embeddings=local_rope if (layer_idx % 6) != 0 or layer_idx == 0 else global_rope,
+            kv_cache=None,
+            max_seq_len=max_seq_len,
+            # QK-norm is required
+            k_norm=GemmaRMSNorm(head_dim, eps=norm_eps),
+            q_norm=GemmaRMSNorm(head_dim, eps=norm_eps),
+            attn_dropout=attn_dropout,
+            # perform global only on the each 6 layer, according to the tech-report
+            sliding_window_size=sliding_window_size if (layer_idx % 6) != 0 or layer_idx == 0 else None,
```
[the remark this answers is not in the document]
Comment (krammnic) [reply torchtune/models/gemma3/_component_builders.py:240]: Let's do it in a follow up

File: torchtune/models/gemma3/_component_builders.py [hunk 3]
```diff
@@ -0,0 +1,392 @@
...
+    Returns:
+        TransformerDecoder: Instantiation of Gemma model with LoRA applied to
+        a subset of the attention projections in each layer.
+    """
+
+    tok_embeddings = GemmaNormEmbeddings(vocab_size, embed_dim)
+    output_proj = TiedLinear(tok_embeddings)
+    
+    local_rope = RotaryPositionalEmbeddings(dim=head_dim, max_seq_len=max_seq_len, base=local_rope_base)
+    global_rope = RotaryPositionalEmbeddings(dim=head_dim, max_seq_len=max_seq_len, base=global_rope_base)
```
[the remark this answers is not in the document]
Comment (krammnic) [reply torchtune/models/gemma3/_component_builders.py:205]: We have done this (according to the tech report):

"We increase RoPE base frequency from 10k to 1M on globa lself-attention layers, and keep the frequency of the local layers at 10k."

But yes, I'm not sure about this:
"We find a scaling factor of 8 to work well in practice." 

Let me investigate little bit on this further.

File: torchtune/models/gemma3/siglip/_component_builders.py [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+                intermediate_dim=intermediate_dim,
+                layer_norm_eps=layer_norm_eps,
+            )
+            for _ in range(num_hidden_layers)
+        )
+
+        self.final_norm = nn.LayerNorm(embed_dim, layer_norm_eps)
+        self.avg_pool = SiglipAveragePooling()
+
+    @torch.inference_mode
```
[the remark this answers is not in the document]
Comment (krammnic) [reply torchtune/models/gemma3/siglip/_component_builders.py:196]: They do not update the SigLIP model on the training and post-training, using same pre-trained model for the all models

File: torchtune/models/gemma3/_convert_weights.py [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+
+"""
+Gemma 3 and Gemma original implementations share different normalization but with
+the same name, so it is mandatory to differentiate their state dict in order to map
+correctly the different weights.
+They are essentially the same except for "model.layers.{}.post_attention_layernorm.weight" key.
+See discussion here: https://github.com/pytorch/torchtune/pull/1835#discussion_r1803410251
+"""
+
+_GEMMA3_FROM_HF = {
```
[the remark this answers is not in the document]
Comment (krammnic) [reply torchtune/models/gemma3/_convert_weights.py:25]: Assuming that a) is a first a) not a second one. The problem is that it looks like we can't do it like this, because of the different structure of the checkpoints :/ Speaking about config.json problem, check this out: https://huggingface.co/google/gemma-3-4b-it/discussions/14

**Comment by krammnic** [comment]
> 1. o the actual builders are all text-only. I think that's fine if there's still stuff up in the air (viz. (2)), but wonder if we should do something similar to our other multimodal models: provide `g`

1.  Sure! I will push this changes (with `EarlyFusion` after we will decide on the structure of the 1B vs 4B+ and discuss the blocker)
2. Speaking about blocker: https://huggingface.co/google/gemma-3-4b-it/discussions/14

**Comment by pocca2048** [comment]
The config issue doesn’t seem likely to get resolved anytime soon and doesn't seem to be a big problem.
How about ignoring it for now and moving forward?

File: torchtune/models/gemma3/_model_builders.py [hunk 6]
```diff
@@ -0,0 +1,363 @@
...
+    from: https://github.com/google/gemma_pytorch/blob/main/gemma/config.py
+
+    Returns:
+        TransformerDecoder: Instantiation of Gemma3 27B model
+    """
+    return gemma3(
+        vocab_size=262_144,
+        num_layers=62,
+        num_heads=32,
+        head_dim=256,
```
Comment (pocca2048) [inline torchtune/models/gemma3/_model_builders.py:291]:
```suggestion
        head_dim=128,
```

File: torchtune/models/gemma3/_model_builders.py [hunk 7]
```diff
@@ -0,0 +1,363 @@
...
+        quantize_base (bool): Whether to quantize base model weights
+
+    Returns:
+        TransformerDecoder: Instantiation of Gemma3 27B model with LoRA applied
+    """
+    return lora_gemma3(
+        vocab_size=262_144,
+        num_layers=62,
+        num_heads=32,
+        head_dim=256,
```
Comment (pocca2048) [inline torchtune/models/gemma3/_model_builders.py:338]:
```suggestion
        head_dim=128,
```

**Comment by krammnic** [comment]
I agree, let's go forward with this @ebsmothers

**Comment by tsvisab** [comment]
Waiting for this, any chance it's going to be merged ? :)

**Comment by krammnic** [comment]
@ebsmothers we can merge without multimodality due to our business with our PRs.

**Comment by krammnic** [comment]
> Waiting for this, any chance it's going to be merged ? :)

I assume we will merge it soon

**Comment by bradhilton** [comment]
Any updated eta on when this may be merged? cc @ebsmothers @krammnic

**Comment by rlrs** [comment]
Interested in getting this merged, willing to help if needed.

**Comment by mistysesame** [comment]
For gemma-3-27b-it, I could not get the logits of this implementation to match the HF version, so I begin to track down where they deviate.
This leads me to a bug in HF-to-torchtune weight conversion [here](https://github.com/krammnic/torchtune/blob/gemma3/torchtune/models/gemma3/_convert_weights.py#L39-L40).
According to the [implementation of `TansformerSelfAttentionLayer`](https://github.com/krammnic/torchtune/blob/gemma3/torchtune/modules/transformer.py#L127-L130), `mlp_norm` is the norm before FF and `mlp_scale` is the norm after FF, but the assignment in the above weight conversion code is the opposite.
To fix this, we should simply swap the two assignment lines: `pre_feedforward_layernorm.weight` should go to `mlp_norm` and `post_feedforward_layernorm.weight` should go to `mlp_scale`.

**Comment by krammnic** [comment]
Let's continue with this. Thanks for the reviews and attention to this PR!

**Comment by rhn19** [comment]
1. gemma3/12B_lora_single_device.yaml & all the other gemma3 12B configs have a typo in checkpoint_files: model-00001-of-00002.safetensors

2. tune run lora_finetune_single_device --config gemma3/12B_lora_single_device throws an exception

torchtune/torchtune/models/convert_weights.py", line 59, in get_mapped_key
    raise Exception(
Exception: Error converting the state dict. Found unexpected key: "language_model.model.embed_tokens.weight". Please make sure you're loading a checkpoint with the right format.

I haven't spent enough time on this library, so please let me know if there's something I'm missing. Happy to provide more information if needed.

**Comment by krammnic** [comment]
I believe that I've fixed lot's of things. Few manual tests and I will merge
